src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionCommitManager.cs

```diff
@@ -0,0 +1,167 @@
+            }
+
+            if (completionService.GetProvider(roslynItem) is IFeaturesCustomCommitCompletionProvider featuresCustomCommitCompletionProvider)
+            {
+                var commitBehavior = CustomCommit(document, featuresCustomCommitCompletionProvider, session.TextView, buffer, roslynItem, session.ApplicableToSpan, typeChar, token);
+                return new EditorCompletion.CommitResult(isHandled: true, commitBehavior);
+            }
+
+            if (document.Project.Language == LanguageNames.VisualBasic && typeChar == '\n')
+            {
+                return new EditorCompletion.CommitResult(isHandled: false, EditorCompletion.CommitBehavior.RaiseFurtherReturnKeyAndTabKeyCommandHandlers);
+            }
+
+            if (item.InsertText.EndsWith(":") && typeChar == ':')
+            {
+                return new EditorCompletion.CommitResult(isHandled: false, EditorCompletion.CommitBehavior.SuppressFurtherTypeCharCommandHandlers);
+            }
+
+            return CommitResultUnhandled;
+        }
+
+        /// <summary>
+        /// This method needs to support custom procesing of commit characters to be on par with the old completion implementation.
+        /// </summary>
+        private bool IsCommitCharacter(char typeChar, ImmutableArray<CharacterSetModificationRule> rules)
+        {
+            // Tab, Enter and Null (call invoke commit) are always a commit character
+            if (typeChar == '\t' || typeChar == '\n' || typeChar == '\0')
+            {
+                return true;
+            }
+
+            foreach (var rule in rules)
+            {
+                switch (rule.Kind)
+                {
+                    case CharacterSetModificationKind.Add:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return true;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Remove:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return false;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Replace:
+                        return rule.Characters.Contains(typeChar);
+                }
+            }
+
+            return s_commitChars.Contains(typeChar);
+        }
+
+        private EditorCompletion.CommitBehavior CustomCommit(
+            Document document,
+            IFeaturesCustomCommitCompletionProvider provider,
+            ITextView view,
+            ITextBuffer buffer,
+            RoslynCompletionItem roslynItem,
+            ITrackingSpan applicableSpan,
+            char commitCharacter,
+            CancellationToken token)
+        {
+            bool includesCommitCharacter;
+            if (!buffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
```

Comment:
It could be good for us but not for users. Not?

---
In reply to: [221381458](https://github.com/dotnet/roslyn/pull/29016#discussion_r221381458) [](ancestors = 221381458)

Reply:
we need to at least assert or NFW here so we can figure out what sort of badness is happening.